evm/p2p/state.py

```diff
@@ -0,0 +1,187 @@
+import asyncio
+import logging
+import random
+import time
+from typing import (  # noqa: F401
+    Any,
+    cast,
+    Dict,
+    List,
+)
+
+import rlp
+
+from trie.sync import HexaryTrieSync
+from trie.exceptions import SyncRequestAlreadyProcessed
+
+from eth_keys import datatypes  # noqa: F401
+from eth_utils import (decode_hex, encode_hex)
+
+from evm.constants import (
+    BLANK_ROOT_HASH,
+    EMPTY_SHA3,
+)
+from evm.rlp.accounts import Account
+from evm.utils.keccak import keccak
+from evm.p2p import eth
+from evm.p2p import protocol
+from evm.p2p.peer import BasePeer, ETHPeer, PeerPool
+from .constants import MAX_STATE_FETCH
+
+
+class StateDownloader:
+    logger = logging.getLogger("evm.p2p.state.StateDownloader")
+    _pending_nodes = {}  # type: Dict[Any, float]
+    _total_processed_nodes = 0
+    _report_interval = 10  # Number of seconds between progress reports.
+    # TODO: Experiment with different timeout/max_pending values to find the combination that
+    # yields the best results.
+    _max_pending = 5 * MAX_STATE_FETCH
+    _reply_timeout = 5  # seconds
+    _last_report_time = time.mktime(time.gmtime(0))
+
+    def __init__(self, chaindb, state_db, root_hash, network_id, privkey):
+        self.peer_pool = PeerPool(ETHPeer, chaindb, network_id, privkey, self.msg_handler)
+        self.root_hash = root_hash
+        self.scheduler = StateSync(root_hash, state_db, self.logger)
+
+    def msg_handler(self, peer: BasePeer, cmd: protocol.Command,
+                    msg: protocol._DecodedMsgType) -> None:
+        """The callback passed to BasePeer, called for every incoming message."""
+        peer = cast(ETHPeer, peer)
+        if isinstance(cmd, eth.NodeData):
+            self.logger.debug("Processing NodeData with %d entries" % len(msg))
+            for node in msg:
+                self._total_processed_nodes += 1
+                node_key = keccak(node)
+                try:
+                    self.scheduler.process([(node_key, node)])
+                except SyncRequestAlreadyProcessed:
+                    # This means we received a node more than once, which can happen when we retry
+                    # after a timeout.
+                    pass
+                # Since a node may be received more than once, check before trying to remove it
+                # from _pending_nodes.
+                if node_key in self._pending_nodes:
+                    self._pending_nodes.pop(node_key)
+
+    # FIXME: Need a better criteria to select peers here.
```

Comment:
I've imagined us storing a local database of peers with somewhat simple metrics attached to them so that we could rank peers by some sort of simple score based on how reliable and fast the peer is.

Reply:
Yep, something like that is what I had in mind as well, but here we might want to rank peers by latency based solely on our interactions with them during the downloading of the state and not from previous interactions as well.

Reply:
Think you understand the problem sufficiently well to write it up in an issue? I can write up something generic but I don't have a good sense for what parts of the code it would fit into.

Reply:
Do you think #282 is a good start?